### DISCUSSION TITLE
LoRa connection between Leader and Tracker

### DISCUSSION THREAD

**Discussion by alphafox02** [nzymedefense/nzyme#343] [Questions]
Noticed the discussion/support group created here and figured it'd be better to reach out with questions here.

I received the correct LoRa devices and went about setting up a leader/tracker for testing. I've ran the python scripts a couple times just to make sure both devices are on the same settings, placed the jumpers back, and started up Nzyme on both ends. Everything starts up and the tracker Aux/Tx lights blink while the leader's Aux/RX lights blink. 

What I don't see happening yet is the Tracker showing up under the Leader's dashboard. Maybe I'm missing a step. I've made sure the 32 character number/letter key is the same on both ends and restarted a couple times. 

I think I'm close.

**Comment by lennartkoopmann** [comment]
Sounds good! Can you post the logs of your leader and tracker?

**Reply by alphafox02** [reply]
Sure, and now that I'm looking at them I don't seem to see anything that's LoRa related. I'm assuming it's in the nzyme.log. 

https://pastebin.com/7u9faMbH

I'll go back and see if there's something to raise the log level.

**Reply by alphafox02** [reply]
Couple things to add, and it's likely something to do with the following.. I had GPSD set to run on boot, and I'm wondering if that was trying to take over the ttyUSB0 (normally my GPS) since I had the /etc/default/gpsd configured. I stopped all of that and rebooted the Pi and saw this mentioned in the nzyme log

04:23:23.574 [groundstation-pings-0] WARN  horse.wtf.nzyme.bandits.trackers.GroundStation - Transmit queue size is unusually large at <9> entries.

I'm thinking I'll have to wait to test more till later. I have the Leader running on a Live boot USB stick and so if I reboot it I'll have to set it all backup.

**Reply by lennartkoopmann** [reply]
Is the transmit queue growing? That sounds like nzyme is sending commands but the LoRa adapter is not sending the messages out. Can you check that the jumpers on the board are set back to the correct position to operate? The boards will just listen for data forever but never send if they are in configuration mode.

**Reply by alphafox02** [reply]
I wasn’t sure how to tell, is all the info located in the same nzyme.log? 

For the jumpers, I removed the m1/ground I believe it was to allow the python scripts to set settings. I then replaced it next to the m0/ground and had the top two on the string to allow usb connection since I had it plugged in via usb. 

I think initially it may have been gpsd causing an issue, which I stopped and disabled, I then checked /dev and found the pi had ttyusb0 and a ttyA something,  can’t recall. I made sure the tracker file referenced ttyusb0. Then on the laptop I also stopped gpsd and found that be unplugging and plugging my back in the device was actually ttyUSB1. I updated the leader file and restarted nzyme on both devices ensuring the 32 character key was the same. 

I can see not long after starting nzyme that both devices led start blinking Tx and Rx respectively, but not necessarily at the same time, in fact the Rx looks to blink before the Tx on the tracker. I’m sure it’s some configuration I’ve not done correctly,  but there also isn’t a lot to configure haha.

**Reply by alphafox02** [reply]
According to this I have it set to mode 0

LoRa mode selection jumpers
short M0, short M1: transmission mode
short M0, open M1: configuration mode
open M0, short M1: WOR mode
open M0, open M1: deep sleep mode
【Note】
Mode 0: Transmission mode, Module transmit data when users send data to UART interface. Wireless receiving is enabled to receive data and send to UART interface when idle.
Mode 1:When it is defined to Transmit, user need to add wakeup codes before transmitting, receiving is same as Mode 0.
Mode 2: Wireless transmit and wireless receive are disabled, users can configure configuration according to #Registers Configuration
Mode 3: Wireless transmit and wireless receive are disabled, module enter deep sleep mode. Module will configure when switching to other modes.

**Comment by alphafox02** [comment]
Testing again - copying and pasting relevant info.

Tracker confirmed ttyUSB0 device, pins placed back as described above. Pwr constant, with Tx/Aux blinking every so often after Nzyme starts, however this is still coming up in the logs

01:13:07.952 [channel-hopper-0] INFO  horse.wtf.nzyme.bandits.trackers.hid.LogHID - Switching channel from [6] to [7].
01:13:08.354 [groundstation-pings-0] WARN  horse.wtf.nzyme.bandits.trackers.GroundStation - Transmit queue size is unusually large at <13> entries.
^C

**Tracker Conf**

# Please consult the nzyme documentation to find out supported tracker devices.
uplink_device {
  type: SX126X_LORA
  parameters: {
    # Set the encryption key to your own 32 character string of letters and numbers.
    # Must be the same on all trackers and nzyme nodes or communication will fail.
    encryption_key: 12345678123456781234567812345678

    # Will usually be /dev/ttyUSB0 except if you have multiple devices connected.
    # You can find out by looking at `dmesg` after plugging it in.
    serial_port: /dev/ttyUSB0
  }
}

Maybe I'm misunderstanding the jumper locations?

**Reply by alphafox02** [reply]
Looking over this documentation,

https://www.waveshare.com/wiki/SX1268_433M_LoRa_HAT

This piece here makes me think something is not the same on each board, but then again why is the transmit queue high?

    1. When wireless configuring, you need to firstly configure Modules address, NETID, Air speed and Key of both modules to same values. For example, if Modules A: Module address is 1, NETID is 1, Air speed is 2.4Kbps and Key is 1. Module B: Module address is 2, NETID is 2, Air speed is 62.5Kbps and Key is 2. If you want to use Module A to configure Module B via wireless network, you need to set Module A to be same as Module B. Then you can use 0xCF 0xCF command to configure Module B via wireless network. 
    2. After configure temporary registers, LoRa modules work by settings of temporary registers. If LoRa modules restart, the settings of temporary are invalid, LoRa module will re-configure registers according to network settings by 0xC1 commands.

**Reply by alphafox02** [reply]
While Nzyme on the LEADER and TRACKER is running I pull the two bottom pins off of the LEADER LoRa device which will immediately make the TRACKER TX/Aux lights stop blinking. Those two Leds will not start blinking again until I put the LEADER's two bottom pins back on. This has to mean there's some sorta of communication happening between the two devices at some level?

**Comment by alphafox02** [comment]
Did you do anything else to your LoRa adapters besides moving and replacing the jumpers to allow for the three scripts to run to set settings. 

I’ll update everything and try again ASAP. There’s so little configuration needed and your instructions are detailed so it’s baffling me as to why the transmit queue continues to grow yet the devices on each end are showing blinking Tx and Rx almost in sync.

**Reply by lennartkoopmann** [reply]
Can you try to run the read scripts and see what they return for both LoRa modules? They are here: https://github.com/lennartkoopmann/nzyme/tree/master/misc/SX126X

Make sure to stop nzyme before running them.

**Reply by alphafox02** [reply]
I had a minute to check both devices. Using the read scripts confirm both are set at Channel 65, Misc Settings 128, and Speed at 35.
